Introduce client::Background

mzabaluev opened the pull request:
A public named type for the background task future, needed to make the signature of the `Service` trait implementation of `Connect` usable with generic executor types. ~~Replaces `ConnectExecutor`~~ The `ConnectExecutor` alias trait uses this type as the parameter for its `TypedExecutor` prerequisite and is made public.

Fixes #44

mzabaluev committed:
Introduce client::Background

A public named type for the background task future, needed to
make the signature of the Service trait implementation of Connect
usable with generic executor types. Replaces ConnectExecutor.

LucioFranco commented on src/client/background.rs:
```diff
@@ -0,0 +1,56 @@
+use std::fmt::{self, Debug};
+use tokio_io::{AsyncRead, AsyncWrite};
+
+use log::error;
```
can we group this with the others up top?

mzabaluev replied:
OK. This one felt special because: 1) it's a macro import; 2) I'd like it to go away eventually (see #45), so I wanted it to stick out like a sore thumb.

LucioFranco commented on src/client/background.rs:
```diff
@@ -0,0 +1,56 @@
+    connection: HyperConnection<T, B>,
+}
+
+impl<T, B> Debug for Background<T, B>
```
nit: can you move this to the bottom, i find the other impl blocks a lot more important and thus deserve to be higher up.

mzabaluev replied:
OK. Personally, I prefer to keep manual `Debug` impls closer to the definition, to notice quicker if any newly added fields are missing from the output.

LucioFranco commented on src/client/connect.rs:
```diff
@@ -28,15 +25,6 @@ pub struct Connect<A, B, C, E> {
 }
 
-/// Executor that will spawn the background connection task.
-pub trait ConnectExecutor<T, B>:
```
I would like to keep this trait as it simplifies, its pretty much a trait alias. We do this in tower-buffer as well. Now that we have the background type we can simplify the TypeExecutor bound which you wanted. I'm also fine with this trait being public.

LucioFranco commented on src/client/connect.rs:
```diff
@@ -70,22 +58,14 @@ pub enum ConnectError<T> {
 where
     C: HttpMakeConnection<A>,
-    B: HttpBody,
+    B: HttpBody + Send + 'static,
```
hmm is there a reason we need this send and static bound now?

mzabaluev replied:
This admittedly was a follow-up commit to restrict the constructors of `Connect` into only producing usable `MakeService` instances. It's not really needed and it is more intrusive than the `Background` change, so it can be dropped.

mzabaluev replied:
However, for usability in generic code, it's better to have the compiler diagnose failures to meet these bounds on the inherent methods, rather than produce a puzzling error message telling only that `Connect` does not implement `Service` without further specifics. I think this should be addressed during the API polishing phase.

mzabaluev replied:
I have removed this commit.

LucioFranco reviewed (changes requested):
Overall looks great to me. Two small formatting comments and I'd like to keep the `ConnectExecutor` trait. Let me know what you think 😄

mzabaluev committed:
Restore ConnectExecutor

Respecified and blanket implemented in terms of client::Background.
Make it a public type as well.

mzabaluev committed:
Addressed minor review comments

As per discussion in tower-rs/tower-hyper#46

mzabaluev committed:
Internalize LiftBody in ConnectExecutor

ConnectExecutor is a utility trait to bind generic type parameters
as supplied by the API user and expose their bounds required
for an executor. Relegating part of it to LiftBody does not make sense.

LucioFranco commented on src/client/connect.rs:
```diff
@@ -29,11 +27,12 @@ pub struct Connect<A, B, C, E> {
 /// Executor that will spawn the background connection task.
-pub trait ConnectExecutor<T, B>:
-    TypedExecutor<MapErr<HyperConnection<T, B>, fn(hyper::Error) -> ()>>
+pub trait ConnectExecutor<T, B>: TypedExecutor<Background<T, LiftBody<B>>>
```
can we punt this `LiftBody` to `Background`?

so this ends up being:

```rust
pub trait ConnectExecutor<T, B>: TypedExecutor<Background<T, B>> 
```

LucioFranco commented on src/client/background.rs:
```diff
@@ -0,0 +1,55 @@
+    T: AsyncRead + AsyncWrite + Send + 'static,
+    B: Payload,
+{
+    connection: HyperConnection<T, B>,
```
related to my other comment, we can put the LiftBody here in place of B which should be nice to hide the fact that we need to even lift the body.

mzabaluev replied:
Done in 291d222.

LucioFranco reviewed (approved):
Very happy with this, one small thing left and I think this is good.

mzabaluev committed:
Internalized LiftBody in client::Background

carllerche commented on src/client/mod.rs:
```diff
@@ -25,11 +25,13 @@
 mod future;
 
-pub use self::connect::{Connect, ConnectError};
+pub use self::background::Background;
```
I don't believe the type needs to be re-exported. It can be "sealed".

mzabaluev replied:
With the public `ConnectExecutor`, the only reason to have `Background` public is for completeness of documentation.

carllerche replied:
The downside is it is now part of the public API, meaning changes to the type == breaking change.

mzabaluev replied:
It was not published in a crates.io release, so it's not too late for #49 to fix this :smile:
